core/src/test/java/com/datastax/oss/driver/internal/core/session/throttling/ConcurrencyLimitingRequestThrottlerTest.java

```diff
@@ -236,7 +239,116 @@ public void should_reject_enqueued_when_closing() {
     throttler.register(request);
 
     // Then
-    assertThatStage(request.started)
+    assertThatStage(request.ended)
         .isFailed(error -> assertThat(error).isInstanceOf(RequestThrottlingException.class));
   }
+
+  @Test
+  public void should_run_throttle_callbacks_concurrently()
+      throws ExecutionException, InterruptedException, TimeoutException {
+    // Given
+
+    // a task is enqueued, which when in onThrottleReady, will stall for 100ms
+    // register() should automatically start onThrottleReady on same thread
+
+    // start a parallel thread
+    CountDownLatch firstRelease = new CountDownLatch(1);
+    MockThrottled first = new MockThrottled(firstRelease);
+    Runnable r =
+        () -> {
+          throttler.register(first);
+          first.ended.toCompletableFuture().thenRun(() -> throttler.signalSuccess(first));
+        };
+    Thread t = new Thread(r);
+    t.start();
+
+    // wait for the registration threads to reach await state
+    first.started.toCompletableFuture().join();
+    assertThatStage(first.started).isDone();
```

Comment:
One possible issue with using `join()` is that it will block until the future completes.  I think if everything works as it should, this is fine, but if for whatever reason a bug is introduced it may just hang tests. 

Was looking around and i saw that `CompletionStateAssert.isSuccess()` will actually do `get(2, TimeUnit.SECONDS)`, so maybe we can just change these two lines to:
```suggestion
    assertThatStage(first.started).isSuccess();
```

We could do then do this anywhere else that `join` is used

Reply:
Added this and also pushed the register() off to a second thread so that we can catch if it blocks without blocking the whole test.